File: packages/react-native-vision-camera/ios/Hybrid Objects/Image Types/HybridFrame.swift
//
//  HybridFrame.swift
//  VisionCamera
//
//  Created by Marc Rousavy on 29.10.25.
//

import AVFoundation
import Foundation
import NitroModules

final class HybridFrame: HybridFrameSpec, NativeFrame, LazyLockableBuffer {
  var sampleBuffer: CMSampleBuffer?
  let metadata: MediaSampleMetadata
  var isLocked: Bool = false
  private var planesCached: [HybridFramePlane]?
  var pixelBuffer: CVPixelBuffer? {
    return sampleBuffer?.imageBuffer
  }

  init(
    buffer: CMSampleBuffer,
    metadata: MediaSampleMetadata
  ) {
    self.sampleBuffer = buffer
    self.metadata = metadata
    self.planesCached = nil
    super.init()
  }

  var memorySize: Int {
    return pixelBuffer?.memorySize ?? 0
  }

  var timestamp: Double {
    return metadata.timestamp.seconds
  }
  var isMirrored: Bool {
    return metadata.isMirrored
  }
  var orientation: CameraOrientation {
    return metadata.orientation
  }

  var width: Double {
    guard let pixelBuffer else {
      return 0
    }
    return Double(CVPixelBufferGetWidth(pixelBuffer))
  }

  var height: Double {
    guard let pixelBuffer else {
      return 0
    }
    return Double(CVPixelBufferGetHeight(pixelBuffer))
  }

  var bytesPerRow: Double {
    guard let pixelBuffer else {
      return 0
    }
    return Double(CVPixelBufferGetBytesPerRow(pixelBuffer))
  }

  var isValid: Bool {
    return sampleBuffer?.isValid ?? false
  }

  var isPlanar: Bool {
    guard let pixelBuffer else {
      return false
    }
    return CVPixelBufferIsPlanar(pixelBuffer)
  }

  var pixelFormat: PixelFormat {
    guard let pixelBuffer else {
      return .unknown
    }
    let osType = CVPixelBufferGetPixelFormatType(pixelBuffer)
    return PixelFormat(osType: osType)
  }

  lazy var cameraIntrinsicMatrix: [Double]? = {
    guard let sampleBuffer else {
      return nil
    }
    guard let matrix = sampleBuffer.getCameraIntrinsicMatrix() else {
      return nil
    }
    return matrix.toDoubleArray()
  }()

  func dispose() {
    self.unlockBuffer()
    try? self.sampleBuffer?.invalidate()
    self.sampleBuffer = nil
    self.planesCached?.forEach { $0.dispose() }
  }

  func getPlanes() throws -> [any HybridFramePlaneSpec] {
    guard isValid else {
      throw RuntimeError.error(withMessage: "This Frame has already been disposed!")
    }
    guard let pixelBuffer else {
      throw RuntimeError.error(withMessage: "This Frame does not contain a Pixel Buffer!")
    }
    if let planesCached {
      // we have planes cached
      return planesCached
    }
    try ensureBufferLocked()
    let planeCount = CVPixelBufferGetPlaneCount(pixelBuffer)
    let planes = (0..<planeCount).map { index in
      HybridFramePlane(buffer: pixelBuffer, planeIndex: index)
    }
    self.planesCached = planes
    return planes
  }

  func getPixelBuffer() throws -> ArrayBuffer {
    guard isValid else {
      throw RuntimeError.error(withMessage: "This Frame has already been disposed!")
    }
    guard let pixelBuffer else {
      throw RuntimeError.error(withMessage: "This Frame does not contain a Pixel Buffer!")
    }
    try ensureBufferLocked()
    return try ArrayBuffer.fromPixelBuffer(pixelBuffer)
  }

  func getNativeBuffer() throws -> NativeBuffer {
    guard let pixelBuffer else {
      throw RuntimeError.error(withMessage: "This Frame does not contain a Pixel Buffer!")
    }
    return pixelBuffer.asNativeBuffer()
  }

  func convertCameraPointToFramePoint(cameraPoint: Point) throws -> Point {
    guard let pixelBuffer else {
      throw RuntimeError.error(withMessage: "This Frame has already been disposed!")
    }
    let matrix = FrameCoordinateSystemConverter.getCameraToFrameMatrix(
      pixelBuffer: pixelBuffer,
      orientation: orientation,
      isMirrored: isMirrored)
    return cameraPoint.applying(matrix)
  }

  func convertFramePointToCameraPoint(framePoint: Point) throws -> Point {
    guard let pixelBuffer else {
      throw RuntimeError.error(withMessage: "This Frame has already been disposed!")
    }
    let matrix = FrameCoordinateSystemConverter.getFrameToCameraMatrix(
      pixelBuffer: pixelBuffer,
      orientation: orientation,
      isMirrored: isMirrored)
    return framePoint.applying(matrix)
  }
}

File: packages/react-native-vision-camera/android/src/main/java/com/margelo/nitro/camera/hybrids/instances/HybridFrame.kt
package com.margelo.nitro.camera.hybrids.instances

import android.graphics.Matrix
import androidx.camera.core.ImageProxy
import com.margelo.nitro.camera.HybridFramePlaneSpec
import com.margelo.nitro.camera.HybridFrameSpec
import com.margelo.nitro.camera.NativeBuffer
import com.margelo.nitro.camera.Orientation
import com.margelo.nitro.camera.PixelFormat
import com.margelo.nitro.camera.Point
import com.margelo.nitro.camera.extensions.DisposableArrayBuffer
import com.margelo.nitro.camera.extensions.convertPoint
import com.margelo.nitro.camera.extensions.getNativeBuffer
import com.margelo.nitro.camera.extensions.getPixelBuffer
import com.margelo.nitro.camera.extensions.mapToArray
import com.margelo.nitro.camera.extensions.pixelFormat
import com.margelo.nitro.camera.public.NativeFrame
import com.margelo.nitro.core.ArrayBuffer

class HybridFrame(
  override val image: ImageProxy,
  override val orientation: Orientation,
  override val isMirrored: Boolean,
) : HybridFrameSpec(),
  NativeFrame {
  override val timestamp: Double
    get() = image.imageInfo.timestamp.toDouble()
  override val isValid: Boolean
    get() {
      try {
        // accessing format on a Closed Image might throw
        image.format
        return true
      } catch (_: Throwable) {
        return false
      }
    }
  override val width: Double
    get() = image.width.toDouble()
  override val height: Double
    get() = image.height.toDouble()
  override val bytesPerRow: Double
    get() {
      val onlyPlane =
        image.planes.singleOrNull()
          ?: return 0.0
      return onlyPlane.rowStride.toDouble()
    }
  override val pixelFormat: PixelFormat
    get() = image.pixelFormat
  override val isPlanar: Boolean
    get() = image.planes.size > 1

  // TODO: Implement `cameraIntrinsicMatrix`
  override val cameraIntrinsicMatrix: DoubleArray?
    get() = null

  override val memorySize: Long
    get() = image.width * image.height * 4L

  override fun dispose() {
    super.dispose()
    planesCached?.forEach { it.dispose() }
    cachedPixelBuffer?.dispose()
    image.close()
  }

  override fun getNativeBuffer(): NativeBuffer {
    return image.getNativeBuffer()
  }

  private var planesCached: Array<HybridFramePlaneSpec>? = null

  override fun getPlanes(): Array<HybridFramePlaneSpec> {
    planesCached?.let {
      // we have planes cached
      return it
    }
    val planes: Array<HybridFramePlaneSpec> = image.planes.mapToArray { HybridFramePlane(it) }
    planesCached = planes
    return planes
  }

  private var cachedPixelBuffer: DisposableArrayBuffer? = null

  override fun getPixelBuffer(): ArrayBuffer {
    cachedPixelBuffer?.let {
      // We already have it cached
      return it.arrayBuffer
    }
    val pixelBuffer = image.getPixelBuffer()
    cachedPixelBuffer = pixelBuffer
    return pixelBuffer.arrayBuffer
  }

  override fun convertCameraPointToFramePoint(cameraPoint: Point): Point {
    val sensorToBuffer = image.imageInfo.sensorToBufferTransformMatrix
    return sensorToBuffer.convertPoint(cameraPoint)
  }

  override fun convertFramePointToCameraPoint(framePoint: Point): Point {
    val bufferToSensor =
      Matrix().apply {
        image.imageInfo.sensorToBufferTransformMatrix.invert(this)
      }
    return bufferToSensor.convertPoint(framePoint)
  }
}
